tests/test_utils.py

```diff
@@ -45,3 +67,94 @@ def test_rsa_key(self) -> None:
+        self.assertFalse(result)
+
+    def test_get_user_id(self) -> None:
+        result = utils.get_user_id()
+        if sys.platform == "win32":
+            self.assertEqual(0, result)
+        else:
+            self.assertNotEqual(0, result)
+
+    @patch("sys.platform", "win32")
+    def test_get_user_id_win32(self) -> None:
+        result = utils.get_user_id()
+        self.assertEqual(0, result)
+
+    @patch("sys.stdout", new_callable=StringIO)
+    @patch("subprocess.Popen", autospec=True)
+    def test_execute_exit_without_error(
+        self, mock_popen: MagicMock, mock_stdout: StringIO
+    ) -> None:
+        process = mock_popen.return_value.__enter__.return_value
+        process.returncode = 0
+        process.communicate.return_value = ("output", "error")
+
+        result = utils.execute("echo", "")
+        self.assertEqual(0, result)
+        self.assertEqual("echo \n", mock_stdout.getvalue())
+        process.communicate.assert_called_once()
+        process.wait.assert_not_called()
+        process.kill.assert_not_called()
+
+    @patch("sys.stdout", new_callable=StringIO)
+    @patch("subprocess.Popen", autospec=True)
+    def test_execute_exit_with_error(
+        self, mock_popen: MagicMock, mock_stdout: StringIO
+    ) -> None:
+        process = mock_popen.return_value.__enter__.return_value
+        process.returncode = 1
+        process.communicate.return_value = ("output", "error")
+
+        self.assertRaises(exceptions.TutorError, utils.execute, "echo", "")
+        self.assertEqual("echo \n", mock_stdout.getvalue())
+        process.communicate.assert_called_once()
+        process.wait.assert_not_called()
+        process.kill.assert_not_called()
+
+    @patch("sys.stdout", new_callable=StringIO)
+    @patch("subprocess.Popen", autospec=True)
+    def test_execute_throw_exception(
+        self, mock_popen: MagicMock, mock_stdout: StringIO
+    ) -> None:
+        process = mock_popen.return_value.__enter__.return_value
+        process.returncode = 1
+        process.communicate.side_effect = Exception("Exception occurred.")
+
+        self.assertRaises(exceptions.TutorError, utils.execute, "echo", "")
+        self.assertEqual("echo \n", mock_stdout.getvalue())
+        process.communicate.assert_called_once()
+        process.wait.assert_called_once()
+        process.kill.assert_called_once()
+
+    @patch("sys.stdout", new_callable=StringIO)
+    @patch("subprocess.Popen", autospec=True)
+    def test_execute_keyboard_interrupt(
+        self, mock_popen: MagicMock, mock_stdout: StringIO
+    ) -> None:
+        process = mock_popen.return_value.__enter__.return_value
+        process.returncode = 1
+        process.communicate.side_effect = KeyboardInterrupt()
+
+        try:
+            utils.execute("echo", "")
+        except KeyboardInterrupt:
+            process.communicate.assert_called_once()
+            process.wait.assert_called_once()
+            process.kill.assert_called_once()
```

Comment:
We must be entering this exception catch block, right? Then, the test should fail if we don't. We should write something similar to:

        with self.assertRaises(KeyboardInterrupt)
            utils.execute("echo", "")
        process.communicate.assert_called_once()
        process.wait.assert_called_once()
        process.kill.assert_called_once()